tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AdoNet/DbScopeFactory.cs

```diff
@@ -0,0 +1,118 @@
+// </copyright>
+
+using System;
+using System.Data;
+using Datadog.Trace.Configuration;
+using Datadog.Trace.ExtensionMethods;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.AdoNet
+{
+    internal static class DbScopeFactory
+    {
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(DbScopeFactory));
+
+        public static Scope CreateDbCommandScope(Tracer tracer, IDbCommand command)
+        {
+            if (TryGetIntegrationDetails(command.GetType().FullName, out var integrationId, out var dbTypeName))
+            {
+                var integration = IntegrationRegistry.GetIntegrationInfo(integrationId.ToString());
+                var operationName = $"{dbTypeName}.query";
+                return CreateDbCommandScope(tracer, command, integration, dbTypeName, operationName);
+            }
+
+            return null;
+        }
+
+        public static Scope CreateDbCommandScope(Tracer tracer, IDbCommand command, IntegrationInfo integration, string dbType, string operationName)
+        {
+            if (!tracer.Settings.IsIntegrationEnabled(integration))
+            {
+                // integration disabled, don't create a scope, skip this span
+                return null;
+            }
+
+            try
+            {
+                Span parent = tracer.ActiveScope?.Span;
+
+                if (parent is { Type: SpanTypes.Sql } &&
+                    parent.GetTag(Tags.DbType) == dbType &&
+                    parent.ResourceName == command.CommandText)
+                {
+                    // we are already instrumenting this,
+                    // don't instrument nested methods that belong to the same stacktrace
+                    // e.g. ExecuteReader() -> ExecuteReader(commandBehavior)
+                    return null;
+                }
+
+                string serviceName = tracer.Settings.GetServiceName(tracer, dbType);
+
+                var tags = new SqlTags
+                           {
+                               DbType = dbType,
+                               InstrumentationName = integration.Name
+                           };
+
+                tags.SetAnalyticsSampleRate(integration, tracer.Settings, enabledWithGlobalSetting: false);
+
+                Scope scope = tracer.StartActiveWithTags(operationName, tags: tags, serviceName: serviceName);
+                scope.Span.AddTagsFromDbCommand(command);
+
+                return scope;
+            }
+            catch (Exception ex)
+            {
+                Log.Error(ex, "Error creating or populating scope.");
+                return null;
+            }
+        }
+
+        public static bool TryGetIntegrationDetails(
+            string commandTypeFullName,
+            out IntegrationIds? integrationId,
+            out string dbType)
+        {
+            if (commandTypeFullName == null || commandTypeFullName.Length < 20)
```

Comment:
We should be careful with this when adding new integrations, I can see it causing frustration when it doesn't work 😄 . Maybe we should have a unit test that explicitly loads all `IAdoNetClientData` types and calls this method to confirm we find them correctly?

Reply:
Good catch, thanks. The check for `< 20` is part a previous optimization attempt that is no longer in the code. I will remove it from now and propose the optimization in a separate PR.

```suggestion
            if (commandTypeFullName == null)
```

Reply:
Removed the entire `if` block in c5a0c11e3c0b3d3bbc1debac7551d4eb0cd9915c.